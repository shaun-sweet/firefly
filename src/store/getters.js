export default {
  user (state) {
    return {
      displayName: state.displayName,
      email: state.email,
      id: state.id
    }
  },

  homesList (state) {
    return Object.keys(state.homes).map(homeId => {
      return {
        label: state.homes[homeId].nickname,
        value: homeId
      }
    })
  },

  selectedHome (state) {
    return state.selectedHome
  },

  isLoadingInitialState (state) {
    return state.appState.isLoadingInitialState
  },

  defaultHome (state) {
    return Object.keys(state.homes).filter(home => state.homes[home].role === 'owner')[0]
  },

  devicesViewList (state, getters) {
    if (!getters.isLoadingInitialState) {
      const devicesViewList = state.homes[state.selectedHome].devicesViewList
      const devicesArray = Object.keys(devicesViewList).map((deviceId) => {
        const device = devicesViewList[deviceId]
        // this is what auto complete fuzzy finder search feature needs to display the properties
        device.stamp = device.primaryStateStatus
        device.label = device.alias
        return device
      })
      return devicesArray.filter((device) => device.export_ui)
    }
    return []
  }

  // notifications (state, getters) {
  //   const selectedHome = state.homes[getters.selectedHome]
  //   return Object.keys(selectedHome.notifications).map(id => selectedHome.notifications[id])
  // },

  // events (state, getters) {
  //   const selectedHome = state.homes[getters.selectedHome]
  //   return Object.keys(selectedHome.events).map(id => selectedHome.events[id])
  // }
}
