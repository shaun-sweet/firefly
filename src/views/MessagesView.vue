<template>
  <q-tabs no-pane-border v-model="selectedTab">
    <q-tab slot="title" name="notifications" label="Notifications" default></q-tab>
    <q-tab slot="title" name="events" label="Events" ></q-tab>
    <div class="main-content">

      <q-tab-pane name="notifications">
        <q-list class="notifications" no-border>
          <q-list-header class="header">
            <h5>Notifications</h5>
          </q-list-header>
          <notification-message
            v-for="notification in notifications"
            :key="notification.id"
            :trigger="notification.message"
            :timestamp="notification.time" 
          />
          </q-list>
      </q-tab-pane>

      <q-tab-pane name="events">
        <q-list class="events" no-border>
          <q-list-header>
            <h5>Events</h5>
          </q-list-header>
          <event-message 
            v-for="event in events"
            :key="event.id"
            :device="event.deviceAlias" 
            :state="getValue(event.event)" 
            :timestamp="event.time" />
          </q-list>
      </q-tab-pane>
    </div>

  </q-tabs>
</template>

<script>
import EventMessage from '@/EventMessage'
import NotificationMessage from '@/NotificationMessage'
import { getMessages } from 'src/firebase'
import get from 'lodash/get'

export default {
  components: {
    EventMessage,
    NotificationMessage
  },
  beforeCreate () {
    const state = this.$store.state
    getMessages(state.selectedHome)
      .then((snaps) => {
        const notificationsObj = snaps[0].val()
        const eventsObj = snaps[1].val()
        this.notifications = Object.keys(notificationsObj).map((id) => {
          notificationsObj[id].id = id
          return notificationsObj[id]
        })
        this.events = Object.keys(eventsObj).map((id) => {
          eventsObj[id].id = id
          eventsObj[id].deviceAlias = get(state, `homes.${state.selectedHome}.devicesViewList.${eventsObj[id].ff_id}.alias`)
          return eventsObj[id]
        })
      })
  },
  methods: {
    getValue (event) {
      var formattedString
      for (let key in event) {
        formattedString = `${key}: ${event[key]}`
      }
      return formattedString
    }
  },
  data () {
    return {
      selectedTab: 'notifications',
      notifications: [],
      events: []
    }
  }
}
</script>
<style lang="stylus" scoped>
.main-content
  padding-right 0.5rem

.notifications .timestamp > .value
  margin-left auto

.header
  margin auto
</style>
