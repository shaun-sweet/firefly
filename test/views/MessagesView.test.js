import { mount, createLocalVue } from 'vue-test-utils'
import Vuex from 'vuex'
import MessagesView from 'src/views/MessagesView.vue'
import getters from 'src/store/getters'
import {initialState, testState} from '../fixtures/state'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('MessagesView', () => {
  let wrapper;
  let store;
  let state;
  let spy;
  let actions;
  beforeEach(() => {
    spy = jest.fn()
    state = {...testState}
    actions = {  subscribeToMessages: spy  }
    store = new Vuex.Store({
      state,
      getters,
      actions
    })
    wrapper = mount(MessagesView,{ store, localVue });
  })
  it('should render the MessagesView correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should call dispatch subscribeToMessages beforeCreate', () => {
    expect(spy).toHaveBeenCalled()
  });

})
