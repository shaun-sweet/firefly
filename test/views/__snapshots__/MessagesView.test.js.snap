// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesView should render the MessagesView correctly 1`] = `
<q-tabs
  no-pane-border=""
>
  <q-tab
    default="default"
    label="Notifications"
    name="notifications"
  />
   
  <q-tab
    label="Events"
    name="events"
  />
   
  <div
    class="main-content"
  >
    <q-tab-pane
      name="notifications"
    >
      <q-list
        class="notifications"
        no-border=""
      >
        <q-list-header
          class="header"
        >
          <h5>
            Notifications
          </h5>
        </q-list-header>
         
        <q-item
          class="row notification"
        >
          <div
            class="device col"
          >
            
          </div>
           
          <div
            class="timestamp col row"
          >
            <span
              class="value"
            >
              
            </span>
          </div>
        </q-item>
        <q-item
          class="row notification"
        >
          <div
            class="device col"
          >
            
          </div>
           
          <div
            class="timestamp col row"
          >
            <span
              class="value"
            >
              
            </span>
          </div>
        </q-item>
      </q-list>
    </q-tab-pane>
     
    <q-tab-pane
      name="events"
    >
      <q-list
        class="events"
        no-border=""
      >
        <q-list-header>
          <h5>
            Events
          </h5>
        </q-list-header>
         
      </q-list>
    </q-tab-pane>
  </div>
</q-tabs>
`;
